src/Microsoft.OpenApi.OData.Reader/Edm/ODataPathProvider.cs

```diff
@@ -360,36 +360,47 @@ private void RetrieveMediaEntityStreamPaths(IEdmEntityType entityType, ODataPath
 
             if (visitedNavigationProperties == null)
             {
-                visitedNavigationProperties = new ();
+                visitedNavigationProperties = new();
             }
-                        
-            var navPropFullyQualifiedName = $"{navigationProperty.DeclaringType.FullTypeName()}/{navigationProperty.Name}";
+
+            string navPropFullyQualifiedName = $"{navigationProperty.DeclaringType.FullTypeName()}/{navigationProperty.Name}";
 
             // Check whether the navigation property has already been navigated in the path
             if (visitedNavigationProperties.Contains(navPropFullyQualifiedName))
             {
                 return;
             }
 
+            // Get the navigation source for this navigation property
+            IEdmVocabularyAnnotatable annotatableNavigationSource = (currentPath.FirstSegment as ODataNavigationSourceSegment).NavigationSource as IEdmVocabularyAnnotatable;
```

Comment:
```suggestion
            IEdmVocabularyAnnotatable annotatableNavigationSource = (currentPath.FirstSegment as ODataNavigationSourceSegment)?.NavigationSource as IEdmVocabularyAnnotatable;
```

Reply:
The `FirstSegment` of the path will **always** be a navigation source segment, no?

Reply:
Functionally, maybe, but somebody would have to read all the upstream code to make sure of that, and a change in the upstream code might have impact here. Semantically if first segment is something else, or null, the app will get a null value here and run into a null ref ex, which should be guarded against.

Reply:
Okay. Got it! 
Consequently, the `NavigationRestrictionType` variable assignments will have to go into a `null` condition check then.

![image](https://user-images.githubusercontent.com/40403681/155360241-9604491f-f2e5-4fda-bd75-a122c17cae45.png)